More matrix factorizations

mtfishman opened the pull request:
This moves logic for selecting the proper MatrixAlgebraKit.jl matrix factorization from the tensor level to the matrix level by introducing a submodule `TensorAlgebra.MatrixAlgebra`. This enables making some of the tensor factorization definitions more compact by generating them in an `@eval` loop. It also introduces new shorthand factorizations `polar` and `orth`, where left or right can be selected with a keyword argument.

lkdvos commented on src/factorizations.jl:
```diff
@@ -1,25 +1,42 @@
-  svd_compact!,
-  svd_trunc!,
-  svd_vals!
+using .MatrixAlgebra: MatrixAlgebra
```
```suggestion
```
This is already included in `TensorAlgebra.jl`.
I think it could be clearer to move all imports to that file, to avoid these kinds of things.

lkdvos commented on src/TensorAlgebra.jl:
```diff
@@ -2,6 +2,8 @@ module TensorAlgebra
 export contract, contract!, eigen, eigvals, lq, left_null, qr, right_null, svd, svdvals
 
+include("MatrixAlgebra.jl")
+using .MatrixAlgebra: MatrixAlgebra
```
```suggestion
```
If we don't plan to import any of the names explicitly, I think this isn't even necessary, defining a module makes it available:
```julia
julia> module Test
       f(x) = x
       end
Main.Test

julia> Test.f(1)
1
```

lkdvos commented on src/factorizations.jl:
```diff
@@ -1,25 +1,42 @@
+  return splitdims(X, axes_X), splitdims(Y, axes_Y)
+end
+
+for (f, f_mat) in (
+  (:qr, :(MatrixAlgebra.qr)),
+  (:lq, :(MatrixAlgebra.lq)),
+  (:left_polar, :(MatrixAlgebra.left_polar)),
+  (:right_polar, :(MatrixAlgebra.right_polar)),
+  (:polar, :(MatrixAlgebra.polar)),
+  (:left_orth, :(MatrixAlgebra.left_orth)),
+  (:right_orth, :(MatrixAlgebra.right_orth)),
+  (:orth, :(MatrixAlgebra.orth)),
+  (:factorize, :(MatrixAlgebra.factorize)),
+)
+  @eval begin
+    function $f(A::AbstractArray, labels_A, labels_codomain, labels_domain; kwargs...)
+      biperm = blockedperm_indexin(Tuple.((labels_A, labels_codomain, labels_domain))...)
+      return $f(A, biperm; kwargs...)
+    end
+    function $f(A::AbstractArray, biperm::BlockedPermutation{2}; kwargs...)
+      return factorize_with($f_mat, A, biperm; kwargs...)
+    end
+  end
+end
```
```suggestion
for f in (:qr, :lq, :left_polar, :right_polar, :polar, :left_orth, :right_orth, :orth, :factorize)
  @eval begin
    function $f(A::AbstractArray, labels_A, labels_codomain, labels_domain; kwargs...)
      biperm = blockedperm_indexin(Tuple.((labels_A, labels_codomain, labels_domain))...)
      return $f(A, biperm; kwargs...)
    end
    function $f(A::AbstractArray, biperm::BlockedPermutation{2}; kwargs...)
      return factorize_with(MatrixAlgebra.$f, A, biperm; kwargs...)
    end
  end
end
```

lkdvos commented on src/factorizations.jl:
```diff
@@ -1,25 +1,42 @@
+using .MatrixAlgebra: MatrixAlgebra
+using MatrixAlgebraKit: MatrixAlgebraKit
+
+function factorize_with(f, A::AbstractArray, biperm::BlockedPermutation{2}; kwargs...)
```
By default this does not specialize on `f`, so this might alter type stability a bit (make it less likely that the constprop of the kwargs that influence the output type work)
Since we're using this in the eval loop anyways, it might make sense to simply copy-paste this there directly instead?

lkdvos commented on src/factorizations.jl:
```diff
@@ -161,11 +213,9 @@ function eigvals(
   A::AbstractArray, biperm::BlockedPermutation{2}; ishermitian=nothing, kwargs...
```
```suggestion
  A::AbstractArray, biperm::BlockedPermutation{2}; kwargs...
```
I think this was discarding the `ishermitian` flag and not passing it on before

lkdvos commented on src/MatrixAlgebra.jl:
```diff
@@ -0,0 +1,143 @@
+module MatrixAlgebra
+
+using LinearAlgebra: LinearAlgebra
+using MatrixAlgebraKit:
+  eig_full,
+  eig_full!,
+  eig_trunc,
+  eig_trunc!,
+  eig_vals,
+  eig_vals!,
+  eigh_full,
+  eigh_full!,
+  eigh_trunc,
+  eigh_trunc!,
+  eigh_vals,
+  eigh_vals!,
+  left_orth,
+  left_orth!,
+  left_polar,
+  left_polar!,
+  lq_full,
+  lq_full!,
+  lq_compact,
+  lq_compact!,
+  qr_full,
+  qr_full!,
+  qr_compact,
+  qr_compact!,
+  right_orth,
+  right_orth!,
+  right_polar,
+  right_polar!,
+  svd_full,
+  svd_full!,
+  svd_compact,
+  svd_compact!,
+  svd_trunc,
+  svd_trunc!
```
```suggestion
using MatrixAlgebraKit
```
Given that we're importing almost everything, and this is contained in a submodule, we could simplify this here

lkdvos commented on src/TensorAlgebra.jl:
```diff
@@ -2,6 +2,8 @@ module TensorAlgebra
 
 export contract, contract!, eigen, eigvals, lq, left_null, qr, right_null, svd, svdvals
```
I didn't get the chance to comment on this before, but I think the export list also needs to be updated

lkdvos reviewed (commented):
Definitely a good idea to have a submodule for this.

I feels a bit odd that there are some functions that exist there and some that do not (eg `svdvals`), but I guess `MatrixAlgebra` is not meant to be used outside of this package and then this does not matter?

mtfishman commented:
@lkdvos thanks for the thoughtful comments, I'll go through them.

mtfishman commented:
> I feels a bit odd that there are some functions that exist there and some that do not (eg `svdvals`), but I guess `MatrixAlgebra` is not meant to be used outside of this package and then this does not matter?

Good catch. I originally left out `svdvals` since it would just be the same definition as `MatrixAlgebraKit.svd_vals`, and I was thinking of `TensorAlgebra.MatrixAlgebra` as new functionality added on top of `MatrixAlgebraKit`. However, for completeness I added it back. And yes, `TensorAlgebra.MatrixAlgebra` right now is meant to mostly be used internally, though I added exports for all of the `TensorAlgebra.MatrixAlgebra` functions (from `TensorAlgebra.MatrixAlgebra`, not from `TensorAlgebra`) which allows us to do `using TensorAlgebra.MatrixAlgebra` and access those functions. I also added tests for them.

lkdvos reviewed (approved):
Overall looks good to me!

Just wanted to share some thoughts I had while reading through this, in the end this is completely subjective and shouldn't have any major implications, but I did want to know how you feel about this:

While the `@eval` loops are nice, there are definitely also some drawbacks associated to them:
- I do feel it becomes more involved to read the source code and immediately capture what is going on. 
- Static analysis tools have a harder time dealing with this, for example Cthulhu combined with VSCode has a harder time mapping type analysis to source code statements (since the source code statements are somewhat hidden).
- It becomes somewhat tricky to debug when you want to insert some print statements.

All of this just to highlight that macros don't come without drawbacks.
Here, in most cases I do think they are warranted and nice, but for example for the "loop"s over `eigen` and `eigen!`, and similarly the other in-place vs non-inplace versions I'm a bit more on the fence: the benefit in terms of convenience might not outweigh the drawbacks. Ultimately I don't think there is a "right" choice, but I don't think it's bad to bring this up and actively consider it every now and again :)

mtfishman commented:
That's fair, I definitely don't like to overuse macros and it is good to assess where it is really worth it. Here I liked it because it guarantees that the logic across definitions are exactly the same. Some of definitions I think are long enough to warrant using `@eval` since I don't want to repeat the same logic multiple times, while as you say others are shorter and maybe don't warrant it, but for consistency I just defined all of them through `@eval` so I didn't have to make a judgement call about where that line is. I also am anticipating we may add more features and options to some of the definitions (particularly `factorize`), and I think having them defined with `@eval` will help to make that easier (as you saw, it is pretty easy to forget to forward the right set of keyword arguments and also it is tough testing every combination of options). Anyway, I'll leave it as it is for now, but I'm happy to reassess that choice going forward.

mtfishman commented:
@ogauthe this will require updating #50 accordingly, but I think that won't be very complicated (if anything it should be easier now since I merged certain definitions together).
